File: src/main/kotlin/com/marcel/baumgardt/model/mapper/OpenMovieDatabaseResponseMapper.java
package com.marcel.baumgardt.model.mapper;

import com.marcel.baumgardt.model.connector.OpenMovieDatabaseRating;
import com.marcel.baumgardt.model.connector.OpenMovieDatabaseResponse;
import com.marcel.baumgardt.model.dto.MovieDetail;
import kotlin.Pair;
import lombok.AllArgsConstructor;
import lombok.Getter;
import org.springframework.stereotype.Component;

import java.util.Arrays;
import java.util.Currency;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@Component
public class OpenMovieDatabaseResponseMapper {

    private final static String ACTORS_SPLIT_OPERATOR = ",";
    private final static String UNKNOWN_CURRENCY_CODE = "UNKNOWN_CURRENCY_CODE";

    public MovieDetail mapToMovieDetail(OpenMovieDatabaseResponse openMovieDatabaseResponse) {
        Pair<String, Long> boxOfficePairCurrencyValue = mapToBoxOfficePairCurrencyValue(openMovieDatabaseResponse.getBoxOffice());
        return new MovieDetail(
                openMovieDatabaseResponse.getTitle(),
                openMovieDatabaseResponse.getWriter(),
                mapToListOfActors(openMovieDatabaseResponse.getActors()),
                mapToAverageRatingScore(openMovieDatabaseResponse.getRatings()),
                Integer.parseInt(openMovieDatabaseResponse.getMetaScore()),
                boxOfficePairCurrencyValue.getSecond(),
                boxOfficePairCurrencyValue.getFirst()
        );
    }


    private Pair<String, Long> mapToBoxOfficePairCurrencyValue(String boxOffice) {
        //TODO FTHW-100 - Upgarde
        String currencySymbol = boxOffice.substring(0, 1);
        String currencyCode = Currency.getAvailableCurrencies().stream()
                .filter(c -> c.getSymbol().equals(currencySymbol))
                .map(Currency::getCurrencyCode)
                .findFirst()
                .orElse(UNKNOWN_CURRENCY_CODE);

        Long value = Long.valueOf(boxOffice.substring(1).replace(",", ""));

        return new Pair<>(currencyCode, value);
    }


    private double mapToAverageRatingScore(List<OpenMovieDatabaseRating> ratings) {
        return ratings.stream()
                .mapToDouble(x -> unifyDoubleValue(x.getValue()))
                .average()
                .orElse(0.0);
    }

    private double unifyDoubleValue(String value) {
        //TODO FTHW-100 - Upgarde
        DoubleType doubleType = DoubleType.getDoubleTypeBySymbol(value);
        switch (doubleType) {
            case PERCENT:
                return Double.parseDouble(value.replace(DoubleType.PERCENT.symbol, "")) / 100.00;
            case SLASH:
                List<Double> values = Arrays.stream(value.split(DoubleType.SLASH.symbol))
                        .map(Double::parseDouble)
                        .collect(Collectors.toList());

                return values.get(0) / values.get(1);
            default:
                return 0.0;
        }
    }

    private Set<String> mapToListOfActors(String actors) {
        return Arrays.stream(actors.split(ACTORS_SPLIT_OPERATOR))
                .map(String::trim)
                .collect(Collectors.toSet());
    }

    @Getter
    @AllArgsConstructor
    enum DoubleType {
        //TODO FTHW-100 - Upgarde
        PERCENT("%"),
        SLASH("/"),
        OTHER("");

        private final String symbol;

        public static DoubleType getDoubleTypeBySymbol(String value) {
            return List.of(PERCENT, SLASH).stream()
                    .filter(x -> value.contains(x.symbol))
                    .findFirst()
                    .orElse(OTHER);
        }
    }
}

File: src/main/kotlin/com/marcel/baumgardt/model/mapper/OpenMovieDatabaseResponseMapper.kt
package com.marcel.baumgardt.model.mapper

import com.google.common.collect.ImmutableList
import com.marcel.baumgardt.model.connector.OpenMovieDatabaseRating
import com.marcel.baumgardt.model.connector.OpenMovieDatabaseResponse
import com.marcel.baumgardt.model.dto.common.MovieDetail
import lombok.AllArgsConstructor
import lombok.Getter
import org.springframework.stereotype.Component
import java.util.*
import java.util.stream.Collectors

@Component
class OpenMovieDatabaseResponseMapper {

    fun mapToMovieDetail(openMovieDatabaseResponse: OpenMovieDatabaseResponse): MovieDetail {
        val (boxOfficeCurrencyCode, boxOfficeCurrencyValue) = mapToBoxOfficePairCurrencyValue(openMovieDatabaseResponse.boxOffice)

        return MovieDetail(
            openMovieDatabaseResponse.title,
            openMovieDatabaseResponse.writer,
            mapToListOfActors(openMovieDatabaseResponse.actors),
            mapToAverageRatingScore(openMovieDatabaseResponse.ratings),
            openMovieDatabaseResponse.metaScore.toInt(),
            boxOfficeCurrencyCode,
            boxOfficeCurrencyValue
        )
    }

    private fun mapToBoxOfficePairCurrencyValue(boxOffice: String): Pair<String, Long> {
        val currencySymbol = boxOffice.substring(0, 1)
        val currencyCode = Currency.getAvailableCurrencies().stream()
            .filter { c: Currency -> c.symbol == currencySymbol }
            .map { obj: Currency -> obj.currencyCode }
            .findFirst()
            .orElse(UNKNOWN_CURRENCY_CODE)
        val value = boxOffice.substring(1).replace(",", "").toLong()
        return Pair(currencyCode, value)
    }

    private fun mapToAverageRatingScore(ratings: List<OpenMovieDatabaseRating>): Double {
        return ratings.stream()
            .mapToDouble { x: OpenMovieDatabaseRating -> unifyDoubleValue(x.value) }
            .average()
            .orElse(0.0)
    }

    private fun unifyDoubleValue(value: String): Double {
        return when (DoubleType.getDoubleTypeBySymbol(value)) {
            DoubleType.PERCENT ->
                getPercentValue(value)
            DoubleType.SLASH -> {
                getSlashValue(value)
            }
            else -> 0.0
        }
    }

    private fun getPercentValue(value: String): Double {
        return value.replace(DoubleType.PERCENT.symbol, "").toDouble() / 100.00
    }

    private fun getSlashValue(value: String): Double {
        val values = Arrays.stream(value.split(DoubleType.SLASH.symbol.toRegex()).toTypedArray())
            .map { s: String -> s.toDouble() }
            .collect(Collectors.toList())
        return values[0] / values[1]
    }

    private fun mapToListOfActors(actors: String): Set<String> {
        return Arrays.stream(actors.split(ACTORS_SPLIT_OPERATOR.toRegex()).toTypedArray())
            .map { obj: String -> obj.trim { it <= ' ' } }
            .collect(Collectors.toSet())
    }

    @Getter
    @AllArgsConstructor
    internal enum class DoubleType(
        val symbol: String
    ) {
        PERCENT("%"),
        SLASH("/"),
        OTHER("");

        companion object {
            fun getDoubleTypeBySymbol(value: String): DoubleType {
                return ImmutableList.of<DoubleType>(PERCENT, SLASH).stream()
                    .filter { x: DoubleType -> value.contains(x.symbol) }
                    .findFirst()
                    .orElse(OTHER)
            }
        }
    }

    companion object {
        private const val ACTORS_SPLIT_OPERATOR = ","
        private const val UNKNOWN_CURRENCY_CODE = "UNKNOWN_CURRENCY_CODE"
    }
}